### ISSUE TITLE
FW_Cache Maximum function nesting level of '256' reached, aborting!

### ISSUE THREAD

**Issue by andreiglingeanu** [ThemeFuse/Unyson#2228]
After @GheorgheP landed #2225 and #2226 - I've got this error popping out. I think that's because we're having a lot more option types going on around in our theme and extensions than you guys actually tested with.

```php
( ! ) Fatal error: Maximum function nesting level of '256' reached, aborting! in /Users/andreiglingeanu/Projects/full/tmp/ct_old_header/wp-content/plugins/unyson/framework/helpers/class-fw-cache.php on line 237
```

And, I don't actually understand why autoload will actually fulfill the `FW_Cache` so much.

<hr>

The pull request #2225 itself looks VERY big and scary to me. I see `FW_Option_Type::register` dropped out from core option types, do you guys changed any semantics of this static method? Does this change even expect some interaction from the users of the framework?

@GheorgheP Can you please describe a bit more the why and how of this pull request? That would be very apreciated, because we (@Creative-Themes) are always working with latest master version of Unyson and we're kinda in a blocking position right now.

cc. @moldcraft

**Comment by andreiglingeanu** [comment]
For the ones looking for a quick fix for that, roll back to [this](3fb4a132bd8e5597ac56a9162c53250d3fbeb70f) commit, at least temporarily.

```bash
git checkout 3fb4a13
```

**Comment by danyj** [comment]
@andreiglingeanu  thnx for heads up ,I had no time to test latest master. 

@moldcraft  @GheorgheP 

should we drop 

    FW_Option_Type::register

from custom option types now?

**Comment by andreiglingeanu** [comment]
@danyj Too early to talk about this.

I've kinda had a chat with @GheorgheP these days about a possible optimisation in the [Builder](https://github.com/ThemeFuse/Unyson-Builder-Extension) but there's nothing so global that comes to my mind, especially a new option type format.

**Comment by andreiglingeanu** [comment]
The problem was from an internal option type that violated the contracts of an option type - it was our mistake.

But anyway, along the way I've tested if unyson is capable of handling a large amount of option types. And, indeed, it does:

```php
class FW_Option_Type_General_Option_Type extends FW_Option_Type
{
    public $type = '';
    public function get_type() { return $this->type; }
    protected function _enqueue_static($id, $option, $data) { }
    protected function _render($id, $option, $data) { return ''; }
    protected function _get_value_from_input($option, $input_value) {
        if (is_null($input_value)) { $input_value = $option['value']; }
        return (string)$input_value;
    }
    protected function _get_defaults() {
        return array(
            'value' => ''
        );
    }
}


for ($i = 0; $i < 1000; $i++) {
    eval("class FW_Option_Type_Test_N$i extends FW_Option_Type_General_Option_Type {
        public \$type = 'fw-test-$i';
    };");

    FW_Option_Type::register('FW_Option_Type_Test_N' . $i);
}
```

But, I'm not sure what are the memory implications of that and there are [some](https://github.com/ThemeFuse/Unyson/issues/2047) problems with `_enqueue_static` to be aware of. But, if you can think of a way to abstract some code into an option type - do this right away, you'll gain lots of benefits from it, especially after @GheorgheP will finish his exploration around optimizing option types.

**Comment by andreiglingeanu** [comment]
> I see FW_Option_Type::register dropped out from core option types, do you guys changed any semantics of this static method? Does this change even expect some interaction from the users of the framework?

That's still a valid question.

**Comment by andreiglingeanu** [comment]
https://github.com/ThemeFuse/Unyson/blob/300e2babe211e64333b05934b7d207958c947be1/framework/includes/hooks.php#L15

Oh, I see. You just introduced one more level of indirection. The thing is that it won't matter anyway to anyone who's not concerned with the framework internals. And yes, I kinda like the fact that the some core classes will be autoloaded at the first reference (option types may not be registered yet, but who cares??).

<hr>

The bottom line is that everything's fine, nothing has changed for themes.
